tensorflow/core/lib/strings/str_util.cc

```diff
@@ -416,6 +417,118 @@ bool ConsumeNonWhitespace(StringPiece* s, StringPiece* val) {
   }
 }
 
+Status SplitUTF8(StringPiece text, const string& delim,
+                 std::vector<string>* result) {
+  // Bytes    Byte 1    Byte 2    Byte 3    Byte 4
+  //   1     0xxxxxxx
+  //   2     110xxxxx  10xxxxxx
+  //   3     1110xxxx  10xxxxxx  10xxxxxx
+  //   4     11110xxx  10xxxxxx  10xxxxxx  10xxxxxx
+  result->clear();
+  result->reserve(text.size());
+  size_t off = 0, len = 0;
+  bool advance = true;
+  for (size_t i = 0; i < text.size(); ++i) {
+    if (off == i) {
+      size_t num_bytes = 0;
+      Status s = UTF8CharNumBytes(text.substr(off, 4), &num_bytes);
+      if (!s.ok()) {
+        result->clear();
+        return s;
+      }
+      if (num_bytes == 1) {
+        string entry = text.substr(off, 1).ToString();
+        if (delim == "") {
+          // If delim is "" then always advance
+          result->emplace_back(entry);
+        } else if (delim != entry) {
+          // If delim is not the current char, then advance only once
+          if (advance) {
+            advance = false;
+            result->emplace_back(entry);
+          } else {
+            (*result)[result->size() - 1] =
+                (*result)[result->size() - 1] + entry;
+          }
+        } else {
+          // If delim is the current char, then set the advance but
+          // not append the string.
+          advance = true;
+        }
+        off++;
+      } else {
+        len = num_bytes;
+      }
+    } else {
+      if (TF_PREDICT_FALSE((text[i] & 0xC0) != 0x80)) {
+        result->clear();
+        return errors::InvalidArgument("Invalid UTF8 encoding at byte of ", i);
+      }
+      if (off + len == i + 1) {
+        string entry = text.substr(off, len).ToString();
+        if (delim == "") {
+          // If delim is "" then always advance
+          result->emplace_back(entry);
+        } else if (delim == entry) {
+          // If delim is the current char, then set the advance but
+          // not append the string.
+          advance = true;
+        } else if (advance) {
+          // If delim is not the current char, then advance only once
+          advance = false;
+          result->emplace_back(entry);
+        } else {
+          (*result)[result->size() - 1] = (*result)[result->size() - 1] + entry;
+        }
+        off += len;
+      }
+    }
+  }
+  if (TF_PREDICT_FALSE(off < text.size())) {
+    result->clear();
+    return errors::InvalidArgument("Not enough characters for UTF8 encoding");
+  }
+  return Status::OK();
+}
+
+Status UTF8CharNumBytes(StringPiece text, size_t* num_bytes) {
+  if ((text[0] & 0x80) == 0x00) {
+    *num_bytes = 1;
+  } else if ((text[0] & 0xE0) == 0xC0) {
+    *num_bytes = 2;
+  } else if ((text[0] & 0xF0) == 0xE0) {
+    *num_bytes = 3;
+  } else if ((text[0] & 0xF8) == 0xF0) {
+    *num_bytes = 4;
+  } else {
+    return errors::InvalidArgument("Invalid UTF8 encoding at position of 0");
```

Comment:
Hmm...this "position 0" isn't adding much value. Perhaps make this function return a `bool` instead and fill in the status message with offset information at the caller?

Reply:
Done.